File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        pointerOne = 0
        pointerTwo = len(heights) - 1
        maxVolume = 0
        while pointerOne < pointerTwo:
            volume = min(heights[pointerOne], heights[pointerTwo]) * (pointerTwo - pointerOne)
            maxVolume = max(maxVolume, volume)

            if heights[pointerOne] < heights[pointerTwo]:
                pointerOne += 1
            else:
                pointerTwo -= 1

        return maxVolume

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-0.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        vector<int> memo(cost.size(), -1);
        
        return min(dp(cost, memo, 0), dp(cost, memo, 1));
    }

    int dp(vector<int>& cost, vector<int>& memo, int i) {
        if (i >= cost.size()) return 0;
        if (memo[i] != -1) return memo[i];

        memo[i] = cost[i] + min(dp(cost, memo, i + 1), dp(cost, memo, i + 2));

        return memo[i];
    }
};
